balena/models/device_os.py

```diff
@@ -259,37 +310,145 @@ def get_device_os_semver_with_variant(self, os_version, os_variant=None):
+
+            if not host_app_device_type:
+                return {}
+            
+            if host_app_device_type not in os_versions_by_device_type:
+                os_versions_by_device_type[host_app_device_type] = []            
+
+            app_tags = self.__get_os_app_tag(host_app['application_tag'])
+            os_versions_by_device_type[host_app_device_type].extend(self.__get_os_versions_from_releases(host_app['owns__release'], app_tags))
+
+        for device_type in os_versions_by_device_type:
+            os_versions_by_device_type[device_type].sort(reverse=True, key=cmp_to_key(sort_version))
+            recommended_per_os_type = {}
+            
+            for version in os_versions_by_device_type[device_type]:
+                if version['os_type'] not in recommended_per_os_type:
+                    if version['variant'] != 'dev' and not version['known_issue_list'] and not version['version']['prerelease']:
+                        additional_format = ' ({line}, recommended)'.format(line=version['line']) if version['line'] else ' (recommended)'
+                        version['is_recommended'] = True
+                        version['formatted_version'] = 'v{version}{additional_format}'.format(version=version['stripped_version'], additional_format=additional_format)
+                        recommended_per_os_type[version['os_type']] = True
+
+        return os_versions_by_device_type
+
     def get_supported_versions(self, device_type):
         """
         Get OS supported versions.
 
         Args:
             device_type (str): device type slug
 
-        Returns:
-            dict: the versions information, of the following structure:
-                * versions - an array of strings, containing exact version numbers supported by the current environment.
-                * recommended - the recommended version, i.e. the most recent version that is _not_ pre-release, can be `None`.
-                * latest - the most recent version, including pre-releases.
-                * default - recommended (if available) or latest otherwise.
-
         """
-
-        response = self.base_request.request(
-            '/device-types/v1/{device_type}/images'.format(device_type=device_type), 'GET',
-            endpoint=self.settings.get('api_endpoint'), auth=False
-        )
-
-        potential_recommended_versions = [i for i in response['versions'] if not re.search(r'(\.|\+|-)dev', i)]
-        potential_recommended_versions = [i for i in potential_recommended_versions if not semver.parse(i)['prerelease']]
-        recommended = potential_recommended_versions[0] if potential_recommended_versions else None
-
-        return {
-            'versions': response['versions'],
-            'recommended': recommended,
-            'latest': response['latest'],
-            'default': recommended if recommended else response['latest']
-        }
```

Comment:
Since there is an user blocked by this and we don't completely change the return object so I marked it as `minor` so it can be released soon but this should be a major.

Reply:
Changed to `major`. Will try to finish other changes asap so this can be released soon.

Reply:
@nghiant2710 any updates on this?

Reply:
@thgreasi, I updated the return value so this PR can be a minor release, will change it again in next major release.

Reply:
can you please review the PR again?